fix for -r on empty directory

brailovich opened the pull request:
fix for -r on empty directory resulted in zstd waiting for input from stdin. this fix makes zstd exit without erroring
but printing a warning message explaining that there were no files or non-empty directories to process.

brailovich committed:
fix for error message in recursive mode for an empty folder

-r on empty directory resulted in zstd waiting input from stdin. now zstd exits without error and prints a warning message explaining why no processing happened (no files or directories to process).

brailovich committed:
Merge pull request #1 from brailovich/brailovich-patch-1

fix for error message in recursive mode for an empty folder

terrelln commented:
Thanks for the PR @brailovich!

I definitely agree with the behavior you want. We definitely don't want to use stdin when it is a TTY. There is a small question of whether we want to use stdin when it isn't a TTY. However, gzip doesn't. So I think that this behavior is fine.

terrelln commented on programs/zstdcli.c:
```diff
@@ -1358,7 +1361,17 @@ int main(int argCount, const char* argv[])
+       was a number of empty directories. In this case
+       stdin and stdout should not be used */
+       if (nbInputFileNames > 0 ){
+        DISPLAYLEVEL(2, "please provide correct input file(s) or non-empty directories -- ignored \n");
```
This should be `DISPLAYLEVEL(1, ` since this is an error.

terrelln commented on programs/zstdcli.c:
```diff
@@ -1358,7 +1361,17 @@ int main(int argCount, const char* argv[])
+       stdin and stdout should not be used */
+       if (nbInputFileNames > 0 ){
+        DISPLAYLEVEL(2, "please provide correct input file(s) or non-empty directories -- ignored \n");
+        CLEAN_RETURN(2);
```
This should also be `CLEAN_RETURN(1);`

Cyan4973 replied:
Actually, if we want the CLI to return without erroring in this scenario, it should be `CLEAN_RETURN(0)`.

Cyan4973 commented:
Agreed with @terrelln comments,
this PR proposes a good behavior for the CLI.

Consider adding a test, to ensure this behavior will be preserved across future changes by future committers.

brailovich commented on programs/zstdcli.c:
```diff
@@ -1358,7 +1361,17 @@ int main(int argCount, const char* argv[])
+       stdin and stdout should not be used */
+       if (nbInputFileNames > 0 ){
+        DISPLAYLEVEL(2, "please provide correct input file(s) or non-empty directories -- ignored \n");
+        CLEAN_RETURN(2);
```
thanks, it should be CLEAN_RETURN(0)

brailovich committed:
Update zstdcli.c

brailovich committed:
Update playTests.sh

brailovich commented:
tests added, return values changed.

Cyan4973 commented on tests/playTests.sh:
```diff
@@ -390,6 +390,18 @@ println "\n===>  recursive mode test "
+# combination of -r with empty folder
+mkdir -p tmpEmptyDir
+zstd -r tmpEmptyDir 2>tmplog2
+if [grep "aborting" tmplog2]; then
```
minor `sh` error here with `[` and `]`

brailovich replied:
thanks, updated!

brailovich committed:
Update playTests.sh

Cyan4973 commented on tests/playTests.sh:
```diff
@@ -390,6 +390,18 @@ println "\n===>  recursive mode test "
+# combination of -r with empty folder
+mkdir -p tmpEmptyDir
+zstd -r tmpEmptyDir 2>tmplog2
+if [ grep "aborting" tmplog2 ]; then
```
Seems this construction is still problematic for `sh` compatibility.

I'm no `sh` expert, and this specific construction is not used in our test script, 
but I can recommend an existing idiom that has worked well so far : 

```
zstd -r tmpEmptydir 2>&1 | grep "aborting" && die "Should not abort on empty directory"
```

This would be equivalent, but you could go for something even simpler : 

```
zstd -r tmpEmptydir 
```

this second test only relies on return code being successful, which, I believe, is the goal of the test ?

brailovich replied:
thanks! my initial thought was to go with just zstd -r tmpEmptydir  but I thought that the test should be more verbose. if this is acceptable, that's the best solution, thanks!

brailovich replied:
btw, "if then" sh syntax allows cleaning tmp dirs/files before exiting.

brailovich committed:
Update playTests.sh

combination of -r with empty folder simplified to comply with sh compatibility tests

Cyan4973 reviewed (approved):
Looks good to me ! Thanks @brailovich for your contribution !
